### ISSUE TITLE
PoolManager.request() with invalid scheme produces KeyError

### ISSUE THREAD

**Issue by haikuginger** [urllib3/urllib3#872]
```
>>> pm = urllib3.PoolManager
>>> pm.request('GET', 'FAKESCHEME://github.com')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.5/site-packages/urllib3/request.py", line 69, in request
    **urlopen_kw)
  File "/usr/local/lib/python3.5/site-packages/urllib3/request.py", line 90, in request_encode_url
    return self.urlopen(method, url, **extra_kw)
  File "/usr/local/lib/python3.5/site-packages/urllib3/poolmanager.py", line 155, in urlopen
    conn = self.connection_from_host(u.host, port=u.port, scheme=u.scheme)
  File "/usr/local/lib/python3.5/site-packages/urllib3/poolmanager.py", line 128, in connection_from_host
    pool = self._new_pool(scheme, host, port)
  File "/usr/local/lib/python3.5/site-packages/urllib3/poolmanager.py", line 87, in _new_pool
    pool_cls = self.pool_classes_by_scheme[scheme]
KeyError: 'FAKESCHEME'
```

We should be fetching by that scheme more resiliently, and raising a more specific error message - possibly a `LocationValueError`?
##

**Comment by shazow** [comment]
I'm open to proposals but I feel the error and stack trace are fairly self-explanatory.

**Comment by haikuginger** [comment]
Fair enough! And we don't have any record of complaints, so it's obviously not affecting too many people.

I think I'll leave this open with the Contributor Friendly tag in case someone wants to take a crack at it.

**Comment by jayvdb** [comment]
Could at least be a unit test, to ensure that this exception does become a different exception in the future...?

**Comment by haikuginger** [comment]
@jayvdb - since we require 100% unit test success, the test addition and code change would need to come in at the same time. 🙂

**Comment by sigmavirus24** [comment]
> since we require 100% unit test success, the test addition and code change would need to come in at the same time. 🙂

Huh? New tests can be added whenever without needing to change code. New tests will not affect the 100% coverage bit or 100% unit test success. @jayvdb wants to ensure the KeyError remains a KeyError, that's the test he's proposing (as much as I don't think we need to actually test this or handle it).

**Comment by haikuginger** [comment]
> @jayvdb wants to ensure the KeyError remains a KeyError, that's the test he's proposing

I think I must have misinterpreted that - I thought he was proposing implementing a test now to check for the changed behavior that might come in later. My mistake.

**Comment by shazow** [comment]
Yea not sure what @jayvdb means. Not sure @sigmavirus24's interpretation is accurate, could you confirm? We could have a `@skipped` test if we want to create a task for the future.

Anyways, I'm not sure we have an agreement on what the proposed improvement would be. We can leave the label on, but the best kinds of contributor-friendly tasks are ones which are clearly-defined (no ambiguity/disagreement about what needs to be done) and reasonably shallow (not a lot of deep-diving into the code).

@haikuginger If someone wanted to pick up this task, what kind of solution would you propose?

**Comment by haikuginger** [comment]
Personally, I think that while reading the stacktrace easily reveals what's going on, we could go one step further and switch the KeyError out for an exception that specifically calls out the scheme as unsupported. 

My main concern is that I don't think there's a clear mental correlation between the action performed and the exception currently raised- if you do `dictionary[key]`, then a KeyError is expected. If you do `pm.get(method, url)`, I don't think that's so much the case.

Catching and raising a custom exception would also give our direct consumers more utility in dealing with user input; if they get a bad or unsupported scheme, they don't necessarily need to parse it themselves and/or look out for a seemingly-odd KeyError; they can catch a `LocationValueError` (or similar) and take appropriate action.

**Comment by shazow** [comment]
I don't disagree in this particular case, but I want to draw attention to the fact that this is true with basically every parameter urllib3 accepts. 

Some examples:

```python
In [3]: http = urllib3.PoolManager('blah')

In [4]: http.request('GET', 'https://www.google.com/')
...
/Users/shazow/projects/urllib3/urllib3/poolmanager.py in connection_from_host(self, host, port, scheme)
    127             # Make a fresh ConnectionPool of the desired type
    128             pool = self._new_pool(scheme, host, port)
--> 129             self.pools[pool_key] = pool
    130
    131         return pool

/Users/shazow/projects/urllib3/urllib3/_collections.py in __setitem__(self, key, value)
     64             # If we didn't evict an existing value, we might have to evict the
     65             # least recently used item from the beginning of the container.
---> 66             if len(self._container) > self._maxsize:
     67                 _key, evicted_value = self._container.popitem(last=False)
     68

TypeError: unorderable types: int() > str()
```

(Incidentally, it only explodes in Py3, not Py2)

Or another one:

```python
n [8]: http.request(42, 'https://www.google.com/')
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-8-f0a079625c35> in <module>()
----> 1 http.request(42, 'https://www.google.com/')

/Users/shazow/projects/urllib3/urllib3/request.py in request(self, method, url, fields, headers, **urlopen_kw)
     62         or even the lowest level :meth:`urlopen`.
     63         """
---> 64         method = method.upper()
     65
     66         if method in self._encode_url_methods:

AttributeError: 'int' object has no attribute 'upper'
```

```python
In [9]: http.request('GET', object())
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-9-b81726aa60b9> in <module>()
----> 1 http.request('GET', object())

...

/Users/shazow/projects/urllib3/urllib3/util/url.py in parse_url(url)
    156
    157     # Scheme
--> 158     if '://' in url:
    159         scheme, url = url.split('://', 1)
    160

TypeError: argument of type 'object' is not iterable
```

Obviously this is a bit of hyperbole, but do we want to make custom self-documenting exceptions for every kind of incorrect input? How do we determine a good heuristic for where to draw the line?

**Comment by haikuginger** [comment]
Obviously, we want to write Python, and not Java or C.

I don't know what an overall heuristic ought to look like. But, in this particular case, we're only getting to this point if we have a URL that "looks valid" (in that it has a properly-formatted scheme), but which has a scheme that we choose not to support (obviously, urllib3 is targeted at HTTP and HTTPS, but that is a choice that we make).

I think there's a difference in how we should fail in cases where we just have awful input (for example, an int as an HTTP method, or a generic object as a URL), as compared to how we should fail when we have *good* input (for example, `ftp://` as the scheme in a valid URL) that we just don't work with.

TL; DR: garbage input unarguably deserves garbage output. Good input we don't work with might deserve a decent explanation.

**Comment by shazow** [comment]
That's fair. :)

Especially if someday we want people to register custom handlers for various schemes, special exceptions for that subsystem makes sense.

**Comment by jayvdb** [comment]
@sigmavirus24 was correct.

**Comment by Lukasa** [comment]
I'd be inclined to say that we should approach this with a YAGNI train of thought. I agree with @shazow that a custom exception makes sense, but it really only makes sense if we allow pluggable schemes. If we don't ever do that, then I don't think that a custom exception is warranted.

**Comment by haikuginger** [comment]
Makes sense. I do think that the recent work around `pool_classes_by_scheme` is a step in that direction; that's why I was even looking at this region of code anyway.

**Comment by pquentin** [comment]
Does that mean that this issue should be closed?

**Comment by SethMichaelLarson** [comment]
Probably. Will reopen if necessary.
